content/1_General/Choosing_Lang.mdx

```diff
@@ -332,9 +332,54 @@ while lo != hi:
 print(-1 if lo == 0 else all_edges[lo-1][0])
 ```
 
-We are not sure whether it is possible to modify the above approach to receive
-full credit (please let us know if you manage to succeed)! Of course, it is
-possible to pass this problem in Python using DSU (a Gold topic):
+It is possible to optimise this approach even further to pass all testcases. This takes around 3.7s to run.
+
+```py
+def main():
+	f = open("wormsort.in","rb")
+	n,m = map(int, f.readline().split())
+	loc = [*map(int, f.readline().split())]
+	edges = [[] for _ in range(n)]
+	weights = []
+	lo,hi = 0, m - 1
+
+	def valid(loc, minW):
+		component = [-1]*n
+		numcomps = 0
+		for i in range(n):
+			if component[i] != component[loc[i] - 1]:
+				return False
+			elif component[i] == -1:
+				todo = [i]
+				component[i] = numcomps
+				for node in todo:
+					for child, weight in edges[node]:
+						if component[child] == -1 and weight >= minW:
+							component[child] = numcomps
+							todo.append(child)
+				numcomps += 1
+		return True
+
+	for line in f:
+		a,b,w = map(int, line.split())
+		edges[a - 1].append((b - 1, w))
+		edges[b - 1].append((a - 1, w))
+		weights.append(w)
+	weights.sort()
+
+	while lo != hi:
+		mid = (lo + hi) // 2
+		if valid(loc, weights[mid]):
+			lo = mid + 1
+		else:
+			hi = mid
+
+	open("wormsort.out","w").write(f"{-1 if lo == 0 else weights[lo-1]}\n")
```

Comment:
can you fix this? (as mentioned in https://forum.usaco.guide/t/general-choosing-a-language-python-solution-for-usaco-silver-wormhole-sort/3699/3?u=benq)

Reply:
Fixed